common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+  //                 \            /
+  //                 mask ^ pattern
+  //                      |
+  // group & MSBs    MSBs - mask
+  //        \            /
+  //    group_MSBs & mask
+  //
+  // This diagram and the operation count are specific to AArch64 where we have
+  // a fused *integer* multiply-add operation.
+  //
+  // While it is superficially similar to the "find zero bytes in a word" bit
+  // math trick, it is different because this is designed to have no false
+  // positives and perfectly produce 0x80 for matching bytes and 0x00 for
+  // non-matching bytes. This is do-able because we constrain to only handle
+  // present matches which only require testing 7 bits and have a particular
+  // layout.
+
+  // Set the high bit of every byte to `1`. The match byte always has this bit
+  // set as well, which ensures the xor below, in addition to zeroing the byte
+  // that matches, also clears the high bit of every byte.
+  uint64_t mask = byte_ints[0] | MSBs;
+  // Broadcast the match byte to all bytes, and mask in the present bits in the
+  // MSBs of each byte. We structure this as a multiply and an add because we
+  // know that the add cannot carry, and this way it can be lowered using
+  // combined multiply-add instructions if available.
+  uint64_t pattern = LSBs * present_byte + MSBs;
+  CARBON_DCHECK(pattern == (LSBs * present_byte | MSBs))
+      << "Unexpected carry from addition!";
+
+  // Xor the broadcast pattern, making matched bytes become zero bytes.
+  mask = mask ^ pattern;
+  // Subtract the mask bytes from `0x80` bytes so that any non-zero mask byte
+  // clears the high bit but zero leaves it intact.
+  mask = MSBs - mask;
+  // Mask down to the high bits, but only those in the original group.
+  mask &= (byte_ints[0] & MSBs);
+
+  CARBON_DCHECK(VerifyRangeMask(mask, [&](uint8_t byte) {
+    return byte == (present_byte | PresentMask);
+  }));
+  return MatchRange(mask);
+}
+
+inline auto MetadataGroup::PortableMatchPresent() const -> MatchRange {
+  if constexpr (Size > 8) {
+    static_assert(Size <= 32, "Sizes larger than 32 not yet supported!");
+    uint32_t mask = 0;
+    uint32_t bit = 1;
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (bytes[i] & PresentMask) {
+        mask |= bit;
+      }
+      bit <<= 1;
+    }
+    return MatchRange(mask);
+  }
+
+  uint64_t mask = byte_ints[0] & MSBs;
+
+  CARBON_DCHECK(VerifyRangeMask(
+      mask, [&](uint8_t byte) { return (byte & PresentMask) != 0; }));
+  return MatchRange(mask);
+}
+
+inline auto MetadataGroup::PortableMatchEmpty() const -> MatchIndex {
+  if constexpr (Size > 8) {
+    static_assert(Size <= 32, "Sizes larger than 32 not yet supported!");
+    uint32_t bit = 1;
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (bytes[i] == Empty) {
+        return MatchIndex(bit);
+      }
+      bit <<= 1;
+    }
+    return MatchIndex(0);
+  }
+
+  // Materialize the group into a word.
```

Comment:
This comment didn't help me.
```suggestion
  // This sets the high bit of every byte unless the corresponding byte of `bytes_ints[0]` is 0. We take advantage of the fact that the bytes in `bytes_ints[0]` are non-zero only if they are either:
  // - present: in which case the high bit of the byte in `bytes_int[0]` will already be set; or
  // - deleted: in which case the byte in `bytes_int[0]` will be 1, and shifting it left by 7 will cause the high bit to be set.
```